### DISCUSSION TITLE
What is the best place to auto-update fields in models? The example of automatic timestamps.

### DISCUSSION THREAD

**Discussion by abournier** [groue/GRDB.swift#1312] [Q&A]
Hello,

Supposing I have a model like this:

```
struct Drink: Identifiable, Codable, TableRecord, FetchableRecord, MutablePersistableRecord {

	var id: String? // UUID
	var name: String

	var createdAt:Date?
	var updatedAt:Date?
}
```

GRDB has now callbacks, so I can populate 1 field such id with its UUID in willInsert():

```
	mutating func willInsert(_ db: Database) throws {
             if id == nil {
			id = UUID().uuidString.lowercased()
		}
	}
```

I checked GRDB code and willInsert() and didInsert() are the uniq callback to be "mutating".
Now, if I want to auto-update the field `updatedAt`, I can't do this in callbacks as other callback methods are non mutating. `updatedAt` should be updated after any insert/update.

I know I can create a custom method to update it before the save, and call it "manually", for example in:

```
func saveDrink(_ drink: inout Drink) throws {		
    // special stuff
    drink.updatedAt = Date()
    // save now
    try dbWriter.write { db in
        try drink.save(db)
    }
}
```

I don't like it at all.

Is-there a way to update this kind of field in `aroundSave() ` or `willSave()` ? If I used these methods with "mutating" keyword, the compiler warns about it. I can remove the warning with the keyword "private". I'm quite sure it's not a good idea too...

Thanks for any suggestion.

**Comment by groue** [comment]
Hello @abournier,

Thanks for having checked the library possibilities before opening an issue 😊 You are totally right, the `PersistableRecord` protocol only allow mutation on inserts, but not on updates or deletions. There's no escape in the current state of the library.

So you're correct when you modify the `updatedAt` property before saving 👍 

The rest of the answer explores the topic, but you won't find any better option there.

---

First, you have another option which is an SQL trigger, but I really not recommend using it:

<details><summary>Trigger (not advised)</summary>

You can instruct SQLite to touch the `updateAt` column of every updates in the `drink` table:

```swift
// Do it only once!
try db.execute(sql: """
    CREATE TRIGGER UpdateDrinkTimestamp UPDATE ON drink BEGIN
        UPDATE drink SET updatedAt = CURRENT_TIMESTAMP WHERE id=old.id;
    END;
    """)
```

What's cool with a trigger is that you can't forget to update the timestamp 👍 Even batch updates are covered.

What's NOT COOL with the trigger is that you can't prevent the timestamp from being updated (not today, and not in the future when you eventually discover that you need this). 😅 

With a trigger, you'd need to call [`saveAndFetch`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/mutablepersistablerecord/saveandfetch(_:onconflict:)) instead of `save`, so that the `drink` Swift struct gets updated with the new timestamp:

```swift
func saveDrink(_ drink: inout Drink) throws {
    try dbWriter.write { db in
        drink = drink.saveAndFetch(db)
    }
}
```

</details>

Also, you can make `Drink` a class instead of a struct. A `Drink` class could update its `updatedAt` timestamp in `willSave` - but you'd lose the immutable safety of Swift structs, so this is also an option I do not recommend.

> I don't like it at all.

You didn't tell why, so I will guess. We're talking about this one single extra line `drink.updatedAt = Date()`, aren't we? Maybe you've used other ORMs in the past, and you're used to built-in support for timestamps? Maybe you worry that it is easy to forget to update the timestamps in other saving methods? I don't know.

What I can tell is that I'm much aware of the feature. I've used ActiveRecord the Ruby ORM for a long time. It ships with built-in support for timestamps.

So... the feature was not ported to GRDB. The main reason is that users of this feature eventually want to perform updates without touching the timestamps. That's how people are. They think they always want to update timestamps, and probably they do, until they discover they don't, because things have changed.

So ActiveRecord supports explicit disabling of timestamp updates:

```ruby
# Explicit
drink.save(touch: false)
```

Then other users discover that they want to disable timestamp updates at a distance. So ActiveRecord supports implicit disabling of timestamp updates:

```ruby
# Implicit (with a 😳global😱)
Drink.record_timestamps = false
drink.a_method_that_calls_save_eventually
Drink.record_timestamps = true
```

Needless to say, a global is not quite concurrency-safe in a multi-threaded application.

Oh, and I did not mention developers who need to inject an altered clock in, say, tests.

It's a never-ending game between convenience and exceptions. Meh.

All in all, I did not spend the time to design the feature correctly, with the necessary amount of flexibility. And it would really not frequently be used on mobile/desktop dev. I admit servers are another story, but GRDB targets apps. Maybe someone will come with a workable design one day?

**Reply by abournier** [reply]
Hello groue,

It seems you're on fire for the new year ! 😊 What a great/long answer !

1) I didn't think about Triggers. 
It's a cool idea but, yes, can be "dangerous" in some cases. So I won't use them for my needs.

2) Setting the date "manually"

>  I don't like it at all.

Sorry. I should have been more explicit.
Yes, I'm coming from the Rails world with the date created/updated fields to auto-update.

1. The fact is I would prefer to have the date update in the model itself, as other fields (createdAt, UUID....) that can be set in willInsert(), and not set in the "saving" methods ( I can avoid some).
2. I will have maybe 15 models with these same fields. Before mass working, I prefer to have the best solution, that's why I asked. ;)
3. My app will have a sync engine. I will have some similar methods to save items, but doing different things on these fields (createdAt, UpdatedAt, lastDeviceChange, lastServerChange).

For example, I have 2 methods to save items from the device and items pulled from the server:

```
// MARK: - SAVE
		func saveDrink(_ drink: inout Drink) throws {
				if drink.name.isEmpty {
						throw DrinkValidationError.missingName
				}
				// special stuff
				drink.manageFieldsBeforeSaveFromDevice()
				// save now
				try dbWriter.write { db in
						try drink.save(db)
				}
		}
```

```
// MARK: - PULL
		func saveDrinkPulled(_ drink: inout Drink) throws {
				// special stuff
				drink.manageFieldsWhenRecordIsPulled()
				// save now
				try dbWriter.write { db in
						try drink.save(db)
				}
		}
```

So currently, I called methods that are set in protocols and adopted by my models.  
Example:

```
protocol CreatedAtUpdatedAtManager {
	
	var createdAt: Date? { set get }
	var updatedAt: Date? { set get }
	
	mutating func manageCreatedAt()
	mutating func manageUpdatedAt()
}

extension CreatedAtUpdatedAtManager {
	
	mutating func manageCreatedAt() {
		if createdAt == nil {
			createdAt = Date()
		}
	}
	mutating func manageUpdatedAt() {
		updatedAt = Date()
	}
	
}
```


```
extension Drink: IdAsUuidManager, UserIdManager, CreatedAtUpdatedAtManager, LastChangeManager {
    // manage fields when saving from the device
    mutating func manageFieldsBeforeSaveFromDevice() {
	 // always update 'updatedAt'
	manageUpdatedAt()
	// always update 'lastDeviceChange'
	manageLastDeviceChange()
    }
...
}
```

Well, my question was more about the updatedAt field only but the way to manage this kind of field could be useful for other model fields.

In all cases, thanks for your answer. I will have some other questions about "best way to do that" rather than "wanting new features". I have studied the CombineDemoApp and many things are working for me in my samml app. However, I wonder sometimes if it's the best way to do things.

**Reply by groue** [reply]
Thanks for sharing your setup @abournier!

My first answer will just tell that I'm currently working on a very similar app, and we share the same techniques.

There is a "database manager", responsible for enforcing database invariants. All writes in the database happen through it (so that it can enforce invariants). Some models (aka records) require some massage before persistence.

The database manager is generally organized as below:

```swift
class AppDatabase {
    // MARK: - Invariant-preserving write methods (aka methods that perform one transaction)

    internal func frobnicatePlayers(withIds ids: [Player.ID]) throws {
        try dbWriter.write { db in
           let players = try Player.fetchAll(db, ids: ids)
           let context = try fetchContext(db)
           for var player in players {
               try frobnicate(db, &player)
               try playerWillSave(db, &player, context: context)
               try player.save(db)
           }
        }
    }

    internal func turnPlayerIntoSmurf(_ player: inout Player) throws {
        try dbWriter.write { db in
           player.isSmurf = true
           let context = try fetchContext(db)
           try playerWillSave(db, &player, context: context)
           try player.save(db)
        }
    }

    // MARK: - Private utilities

    /// The write access
    private let dbWriter: any DatabaseWriter

    /// Do something special with the player
    private func frobnicate(_ db: Database, player: inout Player) throws { ... }

    /// Returns the context necessary for ``playerWillSave(_:player:context:)``.
    private func fetchContext(_ db: Database) throws -> Context { ... }
    
    /// Makes sure timestamps are bumped, or whatever is necessary
    /// before saving a player.
    private func playerWillSave(_ db: Database, player: inout Player, context: Context) throws { ... }
}
```

This is not visible in above sample, but some records in this app share a common protocol, just as in your app (it's not about timestamps, but this is similar). Methods from this protocol are called from methods such as `AppDatabase.playerWillSave`.

As you can see, records are not responsible for their preparation, but the database manager is. In the above sample code, preparing a player requires a context fetched from the database. If each player instance would have to fetch this context, the batch update `frobnicatePlayers(withIds:)` would be very inefficient.

Arguably, ActiveRecord has better support for fat models, thanks to its memory cache. When N players are saved, and `Player#before_save` is called N times, it is no big deal if `Player#before_save` calls `Context.instance`. That's because only one `Context` is actually fetched from the database thanks to the ActiveRecord cache.

---

I'll try to think more about your 15 models, and follow up.

**Reply by abournier** [reply]
> As you can see, records are not responsible for their preparation, but the database manager is.

I completely agreed with this.
So to be organized with multiple models, I have created AppDatabase extensions, each has 1 file. We can consider these files quite as the Rails controllers (CRUD).

**CrudDrink.swift:**

```
extension AppDatabase {
    
    func saveDrink(_ drink: inout Drink) throws {}
    func removeDrink(_ drink: inout Drink) throws {}

}
```

**CrudBook.swift:**

```
extension AppDatabase {
    
    func saveBook(_ book: inout Book) throws {}
    func removeBook(_ book: inout Book) throws {}

}
```

So all my models have their CRUD interface, and of course, the models are not responsible of their creation/update/deletion. Well, its not the initial subject, but I willl call the model timestamps methods from there as shown before.

**Reply by groue** [reply]
All right :-)

So... Did you consider using a base class for your 15 models? After all, there's nothing like ActiveSupport modules or ActiveSupport concerns within the Swift language. All we have in order to share some behavior and allow overrides in specific types are Swift classes.

**Reply by abournier** [reply]
> So... Did you consider using a base class for your 15 models?

Yes, it's an option. My "old" Objective-C/FMDatabase app uses classes. 
But, I want the new app to be more "modern" and to use the structs and its benefits. It should be an advantage in term of concurrency (and also for my future sync engine I think).

However, I'm talking about 10/15 lines to repeat per CrudModel. It's not a big issue after all.
In all cases, thanks for your suggestions and help.

**Reply by DamienPetrilli** [reply]
> And it would really not frequently be used on mobile/desktop dev. I admit servers are another story, but GRDB targets apps.

Hello, 

I am just jumping in because I am working on data syncing across devices. I think this statement is not accurate as the majority of users now expect their data to sync across devices. And knowing when a record was last updated is an important feature to do that: grabbing data since last update, handling conflict when the same record has been updated on two different devices, etc.

> So... the feature was not ported to GRDB. The main reason is that users of this feature eventually want to perform updates without touching the timestamps. That's how people are. They think they always want to update timestamps, and probably they do, until they discover they don't, because things have changed.

Well I would argue that those who use timestamps but do not want it to update all the time didn't design their data correctly. It should be up to them to handle everything manually with a regular date field. None of those cases should be supported.

I am going to look into the above solution, thanks for your help 👍

**Reply by groue** [reply]
> I am just jumping in because I am working on data syncing across devices. I think this statement is not accurate as the majority of users now expect their data to sync across devices.

This is true. I stand corrected, thank you 👍

And I've just started working on an app that requires sync timestamps ;-)

> Well I would argue that those who use timestamps but do not want it to update all the time didn't design their data correctly. It should be up to them to handle everything manually with a regular date field. None of those cases should be supported.

I'm not sure I agree (assuming I understood what you said). If there would exist an automatic timestamping feature that is all or nothing, then consider the use case of adding a transient (non-synced) property to a timestamped record.

All updates to the transient property would touch the timestamp, against the user's will.

The user would have to move the transient property in a separate record/table.

This is bad. One ought not force a fellow developer into [denormalization](https://en.wikipedia.org/wiki/Database_normalization) of the database schema against his will. You never know the problems this will create (refactoring time, migrations, SQL requests to rewrite, performance impacts, extra file size on disk, stress, broken schedules, you name it). My rule of thumb is: a database schema and the data inside has more value than the code that processes it. If something has to be fixed, it is the code, not the schema.

Finally, the user would open an issue in this repository, and rightfully so. I have better purposes for my free time than apologizing for troubles that I was able to foresee. Discussing points of views on technical topics and looking for good solutions with fellow developers is ok, I quite like that 😉

**Comment by groue** [comment]
### Having record instances responsible for their timestamps considered harmful

Several answers in this discussion make individual record instances responsible for their own timestamps.

This idea has merits, but is it acceptable or meaningful that within a single write, not all records have the same timestamp? This is what happens when each individual record accesses "now" in a naive way (`Date()`). Update thousands of records and you can expect significant differences in the value returned by `Date()`, even though all those records are modified in a single shot.

Consider for example this piece of [SQLite documentation](https://www.sqlite.org/lang_datefunc.html):

> The 'now' argument to date and time functions always returns exactly the same value for multiple invocations within the same [sqlite3_step()](https://www.sqlite.org/c3ref/step.html) call.

This means that the following SQL query assigns the same date to all players, even if the update statement takes a lot of time to execute:

```sql
-- Identical timestamps for all players
UPDATE player SET score = score + 1, updateDate = datetime('now')
```

PostgreSQL [is even stricter](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT):

> Since these functions [(`CURRENT_DATE`, ...)] return the start time of the current transaction, their values do not change during the transaction. This is considered a feature: the intent is to allow a single transaction to have a consistent notion of the “current” time, so that multiple modifications within the same transaction bear the same time stamp.

Now, would we want the equivalent Swift code to generate distinct timestamps?

```swift
// Think about what you expect from timestamps here:
for player in players {
    try player.incrementScore(db)
}
```

To me, the clear answer is that we should strive for identical timestamps. I won't assume I'm smarter than SQLite + PostgresQL designers.

A quick Google search for a detailed description of ActiveRecord behavior on this topic did not yield significant answer - but I'm sure Rails was doing the wrong thing a few years ago : https://groups.google.com/g/rubyonrails-talk/c/TtqeqnP9JGI/m/OnZcfJxxQjYJ.

Let's not make the same mistake: record instances must not be responsible for the value of their timestamps.

### Possible solution: database access locals

So, the GRDB user can make sure that several related statements or a whole transaction use the same timestamp.

I'm not sure yet how this could be done.

I'm considering taking inspiration from Swift [`TaskLocal`](https://developer.apple.com/documentation/swift/tasklocal) and introduce "database access locals". A database access local is a value that can be bound and read in the context of a database access. It is available at any point during a database access.

Like `TaskLocal` (I'm lazy), database access locals must be declared as static properties, like this:

```swift
enum DatabaseTimestamp {
    @DatabaseAccessLocal static var currentDate: Date?
}
```

One can access the above database access local with `DatabaseTimestamp.currentDate`:

```swift
extension Player {
    mutating func incrementStore(_ db: Database) {
        try updateChanges(db) {
            $0.score += 1
            // It is a programmer error to forget to set the current date,
            // so shamelessly crash if not set.
            $0.modificationDate = DatabaseTimestamp.currentDate!
        }
    }
}
```

One can set the database access local from within a database access:

```swift
try dbQueue.write { db in
    DatabaseTimestamp.$currentDate.withValue(Date()) {
        for var player in players {
            try player.incrementScore(db)
        }
    }
}
```

### Database access hook

The above implementation of `Player.incrementScore` force unwraps `DatabaseTimestamp.currentDate!` in order to make sure the timestamp has been set before `Player.incrementScore` is called.

I know that the whole concept of "programmer error", and the force unwrap operator, give the shivers to many developers 🤷‍♂️. So we should also make it possible to prevent such programmer error, which means make sure that **all** database accesses set `currentDate`. Let's introduce the "database access hook", a way to execute some code around all database accesses:

```swift
enum DatabaseTimestamp {
    // We can use a non-optional Date now, because
    // we remove the opportunity for programmer errors below
    @DatabaseAccessLocal var currentDate = Date()
}

extension Player {
    mutating func incrementStore(_ db: Database) {
        try updateChanges(db) {
            $0.score += 1
            $0.modificationDate = DatabaseTimestamp.currentDate // 😎
        }
    }
}

// Make sure all database accesses set the current date
var config = Configuration()
config.prepareDatabaseAccess { dbAccess in
    DatabaseTimestamp.$currentDate.withValue(Date(), execute: dbAccess)
}
let dbQueue = try DatabaseQueue("/path/to/db", configuration: config)

// Run!
try dbQueue.write { db in
    for var player in players {
        try player.incrementScore(db)
    }
}
```

I wonder how this `prepareDatabaseAccess` technique works w.r.t dependency injection. Tests, for example, must be able to inject a custom clock.

---

Those are raw thoughts on the topic! What do you think?

**Reply by groue** [reply]
It must be possible to prevent conflicts between two distinct databases:

```swift
try dbQueue1.write { db1 in
    // Here current timestamp was set by dbQueue1's hook.
    try writeStuff(db1)
}

try dbQueue2.write { db2 in
    // Here current timestamp was set by dbQueue2's hook.
    try writeStuff(db2)
}

try dbQueue1.write { db1 in
    writeStuff(db1)
    try dbQueue2.write { db2 in
        // ⚠️ Here we MUST write into dbQueue1 the same timestamps
        // as we did before dbQueue2 access started.
        //
        // So either db1 and db2 use distinct timestamps (impossible
        // with database access locals), either dbQueue2's hook was
        // able to reuse the one set by dbQueue1's hook.
        try writeStuff(db1, db2)
    }
}
```

This shows that the feature is not *trivial*, even with database access locals and database access hooks building blocks. It is too easy to ship an incorrect implementation. This means that the library will have to ship some ready-made timestamp feature (even if implemented on top of locals and hooks).

**Reply by groue** [reply]
> PostgreSQL [is even stricter](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT)

I asked on the SQLite forums: https://sqlite.org/forum/forumpost/c4c7e54ec4

**Reply by DamienPetrilli** [reply]
I like this! The concepts are easy to understand, makes you learn more on Swift itself (`TaskLocal`) so you can apply the knowledge elsewhere. It could also open new capabilities like sharing a transaction ID between transactions as someone suggested on the sqlite forums.

**Reply by groue** [reply]
Totally, Damien :-) My own synchronisation needs are about to involve physical timestamps (`Date`), but also [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamp) that must be incremented only once in a given transaction :-)

**Reply by groue** [reply]
I'm becoming less sure about the "database access hooks". In the case of timestamps (physical or logical), the unit of processing is most likely the transaction, not the database access:

```swift
try dbQueue.writeWithoutTransaction { db in
    // Should both transactions share the same timestamp?
    try db.inTransaction { ... }
    try db.inTransaction { ... }
}
```

So we may have to rely on [`TransactionObserver`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/transactionobserver) instead of introducing the new concept of "database access hook".

**Reply by DamienPetrilli** [reply]
> Totally, Damien :-) My own synchronisation needs are about to involve physical timestamps (`Date`), but also [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamp) that must be incremented only once in a given transaction :-)

If I may: how did you design the database indexes? Do you still keep the internal auto increment primary keys? If so, how do you manage the associations when syncing as the PK are all internal and subjects to conflicts?

I used a `UUID` as a primary key first to solve this issue but then I read about possible performances issues for not using the auto increment PK.

EDIT:
I also tried to look at how CloudKit is doing it. My understanding so far is that they add the UUID to each record but still use the internal PK for associations. But then, when syncing, they map the PK to each UUID and send it to CloudKit (beware I used ChatGPT get the info so not sure if it's accurate). It seems to make sense to keep optimal local performance.

So with the help of PK + UUID + timestamp on each record, they can just sync the delta and keep good perfs

**Reply by abournier** [reply]
> If I may: how did you design the database indexes? Do you still keep the internal auto increment primary keys? If so, how do you manage the associations when syncing as the PK are all internal and subjects to conflicts?

I want to say, maybe it depends your app type and infrastructure.
I keep all PK (and foreign keys) in UUID, app side and server side. I didn't want to manage 1 more field on every platform and the problems added to all nested table relations. If you use both fields (UUID and INT), the complex thing to manage are the relations, except if you don't have ones...

On my app, UUID are strings managed by SQLite. I didn't notice perfs issues but I have only thousands of objects max. I want to say, my uniq perfs issues with SQLite are on complex INNER JOIN.

On the server, I use Postgres, UUID have their own type. I have millions of records with nested records, no perf issues. It's a dedicated server. I added the right indexes at the right places and it works well.

**Reply by DamienPetrilli** [reply]
Thanks for the reply @abournier, that was my first thinking too: have a clean and simple database. But as I am no expert on DB, those PK issues made me think twice.

**Reply by groue** [reply]
Thank you @abournier! Indeed, @DamienPetrilli, UUID primary keys are a great fit for synchronised app that must work offline. If CloudKit has [denormalized](https://en.wikipedia.org/wiki/Denormalization) their schema, then they must have a good reason for this. Until you find *your* good reason for denormalization, you should stick with the simplest schema that fits your needs.

**Reply by groue** [reply]
See https://github.com/groue/GRDB.swift/pull/1331 for the first pull request related to this discussion. It introduces `db.transactioDate`. It returns the start date of the current transaction, and its value does not change during the transaction:

```swift
extension Player {
    mutating func incrementScore(_ db: Database) throws {
        try updateChanges(db) {
            $0.score += 1
            $0.modificationDate = try db.transactionDate // <-
        }
    }
}

try dbQueue.write { db in
    // All players have the same modification date:
    for var player in players {
        try player.incrementScore(db)
    }
}

```

`db.transactionDate` will become the recommended way to deal with modification and creation dates.

As expected, the initial ideas crashed against the wall of reality ([French version](https://twitter.com/bouletcorp/status/410390478526242816)), and the designs described in previous comments were flawed. In particular, each SQLite connection will get its own transaction date. The proposed design matches the [PostgreSQL behavior](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT). I wish SQLite had a similar concept, but it does currently not.

The "database access locals" described in previous comments will likely turn into "database locals" in future pull requests. If the topic of the "current date" can be considered solved, I also want to address needs such as the "current Lamport timestamp". I'll try to find more use cases.

**Reply by DamienPetrilli** [reply]
Looks good! Can't wait to use it!

**Reply by groue** [reply]
v6.7.0 has shipped. See [Record Timestamps and Transaction Date](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/recordtimestamps). I'll spend more time thinking about my own synchronisation needs before I suggest future enhancements - if needed.

**Reply by abournier** [reply]
Thanks @groue for your work ! I will try these new tools soon.

**Reply by groue** [reply]
The DocC article linked above has gained more details about the modification timestamps, but those are not deployed yet. You can read it on GitHub, though: https://github.com/groue/GRDB.swift/blob/master/GRDB/Documentation.docc/RecordTimestamps.md#modification-timestamp

**Reply by groue** [reply]
Since v6.8.0, [Record Timestamps and Transaction Date](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/recordtimestamps) provides a sample protocol `TimestampedRecord` that applications may reuse or take inspiration from.
